server/storage_index.go

```diff
@@ -0,0 +1,504 @@
+// Copyright 2023 The Nakama Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/blugelabs/bluge"
+	"github.com/blugelabs/bluge/index"
+	"github.com/blugelabs/bluge/search"
+	"github.com/gofrs/uuid/v5"
+	"github.com/heroiclabs/nakama-common/api"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+	"time"
+)
+
+type StorageIndex interface {
+	Write(ctx context.Context, objects StorageOpWrites)
+	Delete(ctx context.Context, objects StorageOpDeletes)
+	List(ctx context.Context, indexName, query string, limit int) (*api.StorageObjects, error)
+	Load(ctx context.Context) error
+	Start(runtime *Runtime)
+}
+
+// TODO: Return the modified values so that it can be propagated to other nodes in the cluster
+
+type storageIndex struct {
+	StorageIndexConfig
+	EntryCount *atomic.Uint64
+	Index      *bluge.Writer
+}
+
+type LocalStorageIndex struct {
+	logger                *zap.Logger
+	db                    *sql.DB
+	indexByName           *MapOf[string, *storageIndex]
+	indicesByCollection   *MapOf[string, []*storageIndex] // TODO: Support multiple indices per collection
+	customFilterFunctions *MapOf[string, RuntimeStorageIndexFilterFunction]
+}
+
+func NewLocalStorageIndex(logger *zap.Logger, db *sql.DB, config []StorageIndexConfig) (StorageIndex, error) {
+	blugeCfg := BlugeInMemoryConfig()
+
+	idxByName := &MapOf[string, *storageIndex]{}
```

Comment:
If these are only written once on startup, then let's use a standard `map`.

Reply:
// The MapOf type is optimized for two common use cases: (1) when the entry for a given
// key is only ever written once but read many times, as in caches that only grow,
// or (2) when multiple goroutines read, write, and overwrite entries for disjoint
// sets of keys. In these two cases, use of a MapOf may significantly reduce lock
// contention compared to a Go map paired with a separate Mutex or RWMutex.

I think this usage falls into (1), wdyt?

Reply:
We only ever write to the whole map once, not once per key. The map is filled at startup (no concurrent writes anyway) and then never updated.